src/lib/broadcast.jl

```diff
@@ -45,18 +45,20 @@ function Base.reducedim_init(::typeof(identity), ::typeof(accum), A::AbstractArr
   Base.reducedim_initarray(A, region, nothing, Union{Nothing,eltype(A)})
 end
 
-trim(x, Δ) = reshape(Δ, ntuple(i -> size(Δ, i), Val(ndims(x))))
-trim(x::Tuple, Δ) = NTuple{length(x)}(Δ)
-
-unbroadcast(x::AbstractArray, x̄) =
-  size(x) == size(x̄) ? x̄ :
-  length(x) == length(x̄) ? trim(x, x̄) :
-    trim(x, accum_sum(x̄, dims = ntuple(i -> size(x, i) == 1 ? i : ndims(x̄)+1, Val(ndims(x̄)))))
-
+function unbroadcast(x::AbstractArray, x̄)
+  N = ndims(x̄)
+  if length(x) == length(x̄)
+    _project(x, x̄)  # ProjectTo handles reshape, offsets, structured matrices, row vectors
+  else
+    tup = filter(d -> size(x, d) == 1, ntuple(identity, N))
```

Comment:
```suggestion
    tup = filter(d -> size(x, d) == 1, ntuple(identity, N))  # same as `Tuple(findall(==(1), size(x)))` but carefully optimized
```

Reply:
If it needs a comment, can it be even less optimised? 
```suggestion
    tup = filter(d -> size(x, d) == 1, ntuple(identity, N))  # like findall(size(x) .== 1) but optimised 
```

Reply:
If we didn't need a `Tuple` here then it is much faster if we can do `findall==(1), size(x))`

```julia
julia> const x = rand(3,1,1,2,1);
WARNING: redefinition of constant x. This may fail, cause incorrect answers, or produce other errors.

julia> @btime filter(d -> size(x, d) == 1, ntuple(identity, $(ndims(x))))
  118.460 ns (3 allocations: 80 bytes)
(2, 3, 5)

julia> @btime Tuple(findall(==(1), size($x)))
  154.346 ns (4 allocations: 240 bytes)
(2, 3, 5)

julia> @btime findall(==(1), size($x))
  54.909 ns (3 allocations: 208 bytes)
3-element Vector{Int64}:
 2
 3
 5
```

Reply:
The weird hack of `ndims() + 1` is still the fastest. We could revert to that. My hope was to let more things work with StaticArrays but maybe we settle for not making things worse. 
```
julia> @btime unbroadcast($(rand(1,3)), $(rand(3,3)));  # original
  154.998 ns (5 allocations: 224 bytes)

julia> @btime unbroadcast2($(rand(1,3)), $(rand(3,3)));  # as in PR
  889.191 ns (3 allocations: 112 bytes)

julia> @btime unbroadcast3($(rand(1,3)), $(rand(3,3)));  # with size(x) .== 1
  198.840 ns (8 allocations: 432 bytes)
```

Reply:
lets wrap whatever we come up with in a function like `singleton_dims(x)` or something, so we can encapsulate that complexity away form the implementation of `unbroadcast`

Reply:
For now I've put it back how it was; we can tackle some tidy up on another day.